### PR TITLE
Triggers from multiple RTDB instances

### LINKED ISSUE #2501: Cannot trigger RTDB functions on secondary DB instance when using local emulator
### Description

I posted in https://stackoverflow.com/questions/63145710/how-to-use-a-secondary-db-instance-on-firebase-local-emulator-with-realtime-data but was told to file this as a bug.

I am trying to run Realtime Database triggers on a secondary database instance on the local Firebase Emulator. I have followed the examples in:
* https://stackoverflow.com/questions/59171891/how-do-i-setup-firebase-realtime-db-triggers-for-multiple-databases-using-cloud
* https://firebase.google.com/docs/database/extend-with-functions#specify_the_instance_and_path

by specifying the `instance(dbInstanceId)` after the `functions.database`

But the function will only get triggered on events on the primary database, not on the secondary database.

### [REQUIRED] Environment info

Output of `firebase --version`: 8.6.0

**firebase-tools:**

**package.json**:

```javascript
  "engines": {
    "node": "10"
  },
  "main": "lib/index.js",
  "dependencies": {
    "firebase-admin": "^9.0.0",
    "firebase-functions": "^3.8.0"
  },
```

**Platform:** Windows 10

### [REQUIRED] Test case

Given **lib/index.ts**

```typescript
import * as functions from 'firebase-functions';
import * as admin from 'firebase-admin';

admin.initializeApp();

const defaultDb = admin.database();

// default instance name is 'sgo-personal'
export const onStaffCreatedDefault = functions.database
  .ref('/staff/{name}')
  .onCreate(async (snapshot, context) => {
    defaultDb.ref('/meta/defaultDbCount').set(admin.database.ServerValue.increment(1));
    console.log("Hello from default instance DB");
    return null;
  });

export const onStaffCreatedDb1 = functions.database.instance('db1')
  .ref('/staff/{name}')
  .onCreate(async (snapshot, context) => {
    defaultDb.ref('/meta/db1Count').set(admin.database.ServerValue.increment(1));
    console.log("Hello from DB1");
    return null;
  });

export const onStaffCreatedDb2 = functions.database.instance('db2')
  .ref('/staff/{name}')
  .onCreate(async (snapshot, context) => {
    defaultDb.ref('/meta/db2Count').set(admin.database.ServerValue.increment(1));
    console.log("Hello from DB2");
    return null;
  });
```

### [REQUIRED] Steps to reproduce

1. Add a staff on either `db1` or `db2` instances, no logs are printed.
2. But add a staff in the default instance, all the cloud functions are triggered based on the logs:

```
i  functions: Beginning execution of "onStaffCreatedDefault"
i  functions: Beginning execution of "onStaffCreatedDb2"
>  Hello from default instance DB
i  functions: Finished "onStaffCreatedDefault" in ~1s
i  functions: Beginning execution of "onStaffCreatedDb1"
i  functions: Finished "onStaffCreatedDb1" in ~1s
>  Hello from DB2
>  Hello from DB1
i  functions: Finished "onStaffCreatedDb2" in ~1s
```

### [REQUIRED] Expected behavior

* I expected that `/sgo-personal/meta/db2Count` would be 2 after adding 2 records in `db2`.
* I expected that `/sgo-personal/meta/db1Count` and `/defaultDbCount` would be 1 after adding 1 records each in `db1` and default instance.
* I expected that the "Hello from XXX" logs should only print based on which database instance fired the event.

### [REQUIRED] Actual behavior

* Actual behavior is only events on the default instance triggers **ALL** the functions, when it should have only triggered the `onStaffCreatedDefault` function
* Actual behavior is the `onStaffCreatedDb1` and `onStaffCreatedDb2` are not triggered at all when adding records on those DB instances

See screenshot: Local emulator with 3 DB instances

There's no problem when I deployed them to production. I only encountered this issue when using local emulator. 

I'm looking at https://github.com/firebase/firebase-tools/blob/master/src/emulator/functionsEmulator.ts#L409, which always uses the `projectId` as the namespace instead of the `instanceId`, but I'm not sure if this is the correct starting point to trace the cause of the issue. 

Thanks!

### PR DESCRIPTION
**Pull request by samtstern**
### Description

Fixes #2501, right now all triggers are registered on the default namespace.

### Scenarios Tested

I am now able to add the function trigger to the correct namespace:
```
>>> HTTP REQUEST POST http://localhost:9000/.settings/functionTriggers.json?ns=secondary  
 {"name":"projects/fir-dumpster/locations/_/functions/triggerTwo","path":"/foo/bar","event":"providers/google.firebase.database/eventTypes/ref.write","topic":"projects/fir-dumpster/topics/triggerTwo"}
```

Each function properly triggers:
```
i  functions: Beginning execution of "writeOne"
i  functions: Finished "writeOne" in ~1s
i  functions: Beginning execution of "triggerOne"
>  triggerOne
i  functions: Finished "triggerOne" in ~1s
i  functions: Beginning execution of "writeTwo"
i  functions: Finished "writeTwo" in ~1s
i  functions: Beginning execution of "triggerTwo"
>  triggerTwo
i  functions: Finished "triggerTwo" in ~1s
```

**Functions**
```js
const functions = require('firebase-functions');
const admin = require('firebase-admin');
const { app } = require('firebase-admin');

exports.writeOne = functions.https.onRequest(async (req, res) => {
  if (admin.apps.length === 0) {
    admin.initializeApp();
  }
  
  const db1 = admin.database();
  await db1.ref('/foo/bar').set({
    date: new Date().toISOString(),
  });

  res.send('ok');
});

exports.writeTwo = functions.https.onRequest(async (req, res) => {
  if (admin.apps.length === 0) {
    admin.initializeApp({
      databaseURL: 'https://secondary.firebaseio.com',
      credential: admin.credential.applicationDefault()
    });
  }

  const db2 = admin.database();
  await db2.ref('/foo/bar').set({
    date: new Date().toISOString(),
  });

  res.send('ok');
});

exports.triggerOne = functions.database.ref('/foo/bar').onWrite(() => {
  console.log('triggerOne');
  return true;
});

exports.triggerTwo = functions.database.instance('secondary').ref('/foo/bar').onWrite(() => {
  console.log('triggerTwo');
  return true;
});
```

### Sample Commands

N/A

### CODE REVIEW

**Comment by samtstern** [comment]
@IanWyszynski can you take a look at this and see if it's possibly a bug in the RTDB emulator?

File: src/emulator/functionsEmulator.ts [hunk 1]
```diff
@@ -395,18 +403,19 @@ export class FunctionsEmulator implements EmulatorInstance {
       return Promise.reject();
     }
 
+    const instance = result[1];
     const bundle = JSON.stringify({
       name: `projects/${projectId}/locations/_/functions/${definition.name}`,
```
Comment (IanWyszynski) [inline src/emulator/functionsEmulator.ts:408]: I'm pretty sure we store trigger payloads by unique name.

https://github.com/firebase/firebase-server/blob/master/firebase-server/firebase-core/src/main/scala/com/firebase/core/namespace/NamespaceActor.scala#L1559-L1560

 I'm not exactly sure how `definition` is determined. Will the `name` member be distinct for two different instances?

Comment (samtstern) [reply src/emulator/functionsEmulator.ts:408]: Yeah function names are always unique, there's no way for me to export two functions with the same name but pointing to different instances.  So in this example above the names would be `triggerOne` and `triggerTwo`

File: src/emulator/functionsEmulator.ts [hunk 2]
```diff
@@ -395,18 +403,19 @@ export class FunctionsEmulator implements EmulatorInstance {
...
     });
 
-    logger.debug(`addDatabaseTrigger`, JSON.stringify(bundle));
+    logger.debug(`addDatabaseTrigger[${instance}]`, JSON.stringify(bundle));
 
     let setTriggersPath = "/.settings/functionTriggers.json";
-    if (projectId !== "") {
-      setTriggersPath += `?ns=${projectId}`;
+    if (instance !== "") {
+      setTriggersPath += `?ns=${instance}`;
```
[the remark this answers is not in the document]
Comment (samtstern) [reply src/emulator/functionsEmulator.ts:418]: We fall back to `Constants.DEFAULT_DATABASE_EMULATOR_NAMESPACE` which is defined as `fake-server`.  This behavior is unchanged in this PR, but thanks for checking!

**Comment by samtstern** [comment]
@IanWyszynski ping me if you want to talk about the errors you're having with funcitons.

**Comment by IanWyszynski** [comment]
> @IanWyszynski can you take a look at this and see if it's possibly a bug in the RTDB emulator?

I ran this example with a custom emulator build. I performed:

```
curl -X PUT -d '{"data":true}' http://localhost:9000/foo/bar.json?ns=secondary
```

And:

```
curl -X PUT -d '{"data":true}' http://localhost:9000/foo/bar.json?ns=jw-emu
```

The resulting requests sent by the database emulator to the functions emulator are:

```
POST /functions/projects/secondary/triggers/triggerTwo HTTP/1.1
Content-Length: 307
Host: localhost:5002
Content-Type: application/json; charset=utf-8
```

And

```
POST /functions/projects/jw-emu/triggers/triggerOne HTTP/1.1
Content-Length: 304
Host: localhost:5002
Content-Type: application/json; charset=utf-8
```

From the perspective of the database emulator, the two instances are separate "projects" (in the functions emulator world). The functions emulator, however, assumes that both instances belong to a single project:

https://github.com/firebase/firebase-tools/blob/9d916eac1af98772c61055355300205af6a9e433/src/emulator/functionsEmulator.ts#L161

When we resolve function payloads in prod, it's done on a per-namespace basis. I think the functions emulator should register separate background function routes for each instance in `functions/index.js` so that it recognizes the request to path:

```
/functions/projects/secondary/triggers/triggerTwo
```

**Comment by samtstern** [comment]
@IanWyszynski thank you for investigating!  I just wanted to make sure that it was intended to work this way and that treating namespaces as projects was not a bug.   I will register the extra routes in the functions emulator.

**Comment by samtstern** [comment]
@IanWyszynski bump on this.  It all works now, mind approving?

**Pull request merged**